### ISSUE TITLE
Programs: Search database

### ISSUE THREAD

**Issue by CarlosNZ** [openmsupply/open-msupply#1070]
From: https://github.com/openmsupply/open-msupply/issues/644#issue-1383248595

> #### `Mother`:
> 
>   - only when patient is an infant (Poruan checking the age threshold, most likely expressed in months. We are considering creating a specific program for infants)
>  - the objective is to link the patient record to the mother's patient record. Would it be possible to select mother's patient ID from a dropdown ?

This will be a variation on the standard "Autocomplete" JSON Forms component, but with a database lookup for its source. 

It also needs to:
- only show if patient is infant
- collect all relevant "Mother" info (probably IDs and name)
- maybe a hyperlink to the actual Mother record

I have a suggestion for how to modify existing components to allow more dynamic logic (i.e. database lookups), but I'll need to sit down with someone to clarify the requirements and nut out the details. (@clemens-msupply ?)

**Comment by clemens-msupply** [comment]
There are currently two data sources I can think of:

- Patient registration form (next of kin)
- Other patients in the system (name table)

**Comment by CarlosNZ** [comment]
Just thinking about the requirements for this component, and I think there are two things that distinguish it from a standard "Autocomplete" component:

1. a database lookup to get the list of options
2. the stored data needs to be different to the displayed options (i.e the drop-down would show NUIC / Name but the stored result would need to be an object containing `{id, nuic, name, etc...}`

The first one could be handled with a [FigTree](https://github.com/CarlosNZ/fig-tree-evaluator) expression (as discussed [here](https://github.com/openmsupply/open-msupply/issues/1122), but could also be added as a hard-coded list of available queries, and the config just supplies the parameter.

The second one would need to be an extension to Autocomplete (a "dataMap" option, say?), but that would be easy to add, I think.

If we make it a separate "Search" component, I think it should mostly just be a wrapper for existing "Autocomplete" with those dynamic elements passed in (not exactly sure of the details, but that's the basic idea).

**Comment by CarlosNZ** [comment]
Actually, thinking about this some more, I think it will need to be a separate component, since we don't want it to be an "Autocomplete" -- we don't want the full list of patients to be available (too long, and also privacy issues). So it'll need to be a "search as you type" thing with a minimum number of chars before it starts searching (e.g. 3).

We should have a few (initially 1) built-in queries, such as "Patient lookup", and each query would have the following properties:
- query method (using React Query) (fetches several/most fields, using an input parameter/filter)
- function to map the returned results into display strings
- function to map the selected result label
- list of fields to save for the "data"

Something like this:
```ts
patient: {
  query: (searchParam) => useQuery(searchParam) // where "useQuery" is whatever method "usePatient" provides
  optionDisplayFunction: (patient) => `${patient.code} - ${patient.firstName} ${patient.lastName}`
  labelDisplayFunction: (patient) => `${patient.code} - ${patient.firstName} ${patient.lastName}`
  saveFields: [ "id", "code", "firstName", "lastName" ]
}
```

Then the "options" for this component just needs to have a "query" option, which in this case would be "patient" and the component automatically uses all the data from the above to perform its functionality.

**Comment by clemens-msupply** [comment]
Yes agree with your last comment! I had a config option like this in mind (open for changes):

```typescript
{
  nameLookup?: {
    query?: {
       // e.g. query names in local store or globally 
       source: 'Store' | 'Global' | 'None'
    },
    /**  Lookup name (must be a Person) by path from the current document.  */
    currentDocument?: string[],
    /** Lookup name by path from the patient document */
    patientDocument?: string[],
    saveFields: string[],
  }
}
```

e.g. 
```ts
{
  nameLookup: {
     query: { source: 'Store' },
     patientDocument: ['contactDetails'],
     saveFields: ["id", "code", "firstName", "lastName"]
  }
}
```

**Comment by CarlosNZ** [comment]
Closed by #1223
